### DISCUSSION TITLE
new-session, multiple environment variables (scripting)

### DISCUSSION THREAD

**Discussion by vlappa** [tmux/tmux#3928] [General]
Currently one can set multiple environment variables with `new-session` by using `-e env_var_one=foo, -e, env_var_two=bar` multiple times.  

I would like the user to be able to set the desired environment variables in a config file and then exec a tmux new-session command. So we don't know how many environment variables we need for our command to exec from our application. Such a exec command would look like:  
`"tmux", "new-session", "-e", "env_var_one=foo", "-e", "env_var_two=bar"`

Wouldn't it be more easy to generate a tmux command to exec from a external application, if a alternative syntax would be allowed?   

`new-session -e env_var_one=foo;env_var_two=bar`

`"tmux", "new-session", "-e", "env_var_one=foo;env_var_two=bar"`  

So a string with `;` as delimiter, or a other char. Examples of such a syntax can be found in the filemanager `nnn` and in `fzf`.

**Comment by nicm** [comment]
I'm OK with this if you want to add it but I'd use `:` not `;` because that is more usual. Also need to be careful about `:` inside quotes.

**Reply by vlappa** [reply]
Thx for your reply. I don't see myself coding in C atm, but who knows at some point.

**Reply by vlappa** [reply]
`:` is used in urls, which are used in environment variables, not sure if that's the best choice.

**Comment by krader1961** [comment]
Any punctuation could be included in an env var value. Note that a directory path can also contain colons and semicolons. Which is why you need to define an encoding that provides for disambiguating whether the character is part of a env var value or separates two env vars. Note that the `PATH` env var does not have any encoding as such. So if you do `mkdir a:b` then `export PATH="$PATH:./a:b"` the result won't work as you intended. Not defining an unambiguous encoding has been a mistake made many, many, times. Such as the CSV format popularized by Windows long ago.

**Reply by vlappa** [reply]
Sure, very hard to pick a good one, if that's possible at all for env vars.

**Reply by aless3** [reply]
something like a double colon '::'?
and if it only has a colon treat is as part of the string like any other char obviously 
this way surely it's less "user friendly" - twice as many key - I don't know how many environment vars use double colon as a separator, but I think not much

else double 'something' like double semi-colon or the like? would that be feasible?

**Reply by krader1961** [reply]
> something like a double colon '::'?

No, that doesn't solve the problem. What If I create a directory like this: `mkdir a::b`. Then add that directory to an env var? Since env vars are simply `key=value` pairs where the key and value are arbitrary strings there is literally no single character, or sequence of characters, you can "reserve" as a delimiter that will never appear in either an env var name or value. In fact, the equal-sign, `=` can safely appear in the value since splitting the key and value is done when the first `=` character is seen. Unless, of course, you allow multiple env var definitions in a single string. 

The `PATH` env var obviously suffers from this problem as I mentioned earlier. People who use UNIX like operating systems simply learn to not create directories containing colons to avoid the problem. But that is not an ideal solution. We should strive to avoid repeating the mistakes of the past.
